src/ciflow-bot.ts

```diff
@@ -0,0 +1,199 @@
+import * as probot from 'probot';
+
+export default function runCIFlowBot(app: probot.Application): void {
+  new CIFlowBot(app).main();
+}
+
+export class CIFlowBot {
+  // Constructor required
+  readonly app: probot.Application;
+
+  // Configurations
+  readonly allowed_command: string[] = ['ciflow'];
+  readonly bot_app_name = 'pytorchbot';
+  readonly bot_assignee = 'pytorchbot';
+  readonly pr_label_prefix = 'ciflow/';
+  readonly rollout_users = ['zhouzhuojie']; // slow rollout to specific group of users first
+
+  // Stateful instance variables
+  command = '';
+  command_args: string[] = [];
+  comment_author = '';
+  comment_body = '';
+  ctx: probot.Context = null;
+  dispatch_labels: string[] = [];
+  dispatch_strategies: string[] = ['add_default_labels'];
+  event = '';
+  owner = '';
+  pr_author = '';
+  pr_labels: string[] = [];
+  pr_number = 0;
+  repo = '';
+
+  constructor(app: probot.Application) {
+    this.app = app;
+  }
+
+  valid(): boolean {
+    if (this.event !== 'pull_request' && this.event !== 'issue_comment') {
+      this.ctx.log.error({ctx: this.ctx}, 'Unknown webhook event');
+      return false;
+    }
+
+    if (this.event === 'issue_comment') {
+      if (this.comment_author === '') {
+        this.ctx.log.error({ctx: this.ctx}, 'Empty comment author');
+        return false;
+      }
+
+      // only the pr author can trigger new changes to the ciflow
+      if (this.comment_author !== this.pr_author) {
+        return false;
+      }
+
+      if (!this.allowed_command.includes(this.command)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  rollout(): boolean {
+    if (this.rollout_users.includes(this.pr_author)) {
+      return true;
+    }
+    return false;
+  }
+
+  async dispatch(): Promise<void> {
+    // Default dispatch algorithm: 'add_default_labels'
+    // Just make sure the we add a 'ciflow/default' to the existing set of pr_labels
+    if (this.dispatch_strategies.includes('add_default_labels')) {
```

Comment:
We would want to be picking a single dispatcher, right? This is always going to be `true`, it seems like this might make more sense as a map of strategy -> function that implements it and we call that by a key here / error if it's not there

Reply:
yes, updated to map to a pipeline of functions